app/os_android.go

```diff
@@ -951,3 +1014,45 @@ func Java_org_gioui_Gio_scheduleMainFuncs(env *C.JNIEnv, cls C.jclass) {
 }
 
 func (_ ViewEvent) ImplementsEvent() {}
+
+// start a foreground service
+func Java_org_gioui_StartForeground(id int, title, text string) (chan struct{}, error) {
+	closeChan := make(chan struct{})
+	errChan := make(chan error)
+
+	// get a handle on the startService, stopService methods of GioForegroundService
+	// run everything in a goroutine so that start/stop calls are from the same thread
+	go func() {
+		runInJVM(javaVM(), func(env *C.JNIEnv) {
+			// create an intent and set it to the foregroundService
+			intent, err := getIntent(env, android.appCtx, foregroundService)
+			if err != nil {
+				errChan <- err
+				return
+			}
+
+			// set the notification text
+			setforegroundIntentExtras(env, intent, id, title, text)
+
+			// locate the methods to start and stop our service
+			cls := getObjectClass(env, android.appCtx)
+			startServiceMethod := getMethodID(env, cls, "startService", "(Landroid/content/Intent;)Landroid/content/ComponentName;")
+			stopServiceMethod := getMethodID(env, cls, "stopService", "(Landroid/content/Intent;)Z")
+
+			// Call startService with the above intent
+			_, err = callObjectMethod(env, android.appCtx, startServiceMethod, jvalue(intent))
+			if err != nil {
+				errChan <- err
+				return
+			}
+
+			// wait for the channel to close, and then halt the foreground service
+			// create an intent and set it to the foregroundService
+			errChan <- nil
+			<-closeChan
+			callVoidMethod(env, android.appCtx, stopServiceMethod, jvalue(intent))
```

Comment:
This is subtly incorrect. If there are more than one pending StartForeground, the first closeChan closed would stop the service even though others are still doing foreground work.

I suggest starting a singleton foreground goroutine with once.Do, and make it accept startForeground, stopForeground commands. It will then only stop GioForegroundService if the number of outstanding foreground commands is 0.